### PR TITLE
Use C++ template types in code generation

### PR DESCRIPTION
**Pull request by jsmall-nvidia**
C++ code generation previously generated classes as needed. This style had the advantage that the same mechanism could be used for C++ and C. The downside is that the prelude could not hold any definitions of vector/matrix types which meant such types need to be defined on the client side before calling into the generated shared library.

In this change the generated code uses the templates for those types. 

Note that because partial specialization is used - it requires a newer version of C++ (C++14), which maybe an issue.

### CODE REVIEW

**Comment by tfoleyNV** [comment]
Is this awaiting review?

Also: I'm confused by the comment saying that C++14 is required because of partial specialization. I thought partial template specialization was present in C++ from quite a while ago. I guess I'm also not sure why it is required, but I suppose I'll find out when I review the changes.

**Comment by jsmall-nvidia** [comment]
I'm a little in two mind about it - I won't go into it. For now it's probably fine to review and if appropirate merge. 

On C++14, I'm not 100% sure - I found two references that claimed partial specialization is a C++14 feature.  Here's one of them https://en.cppreference.com/w/cpp/language/partial_specialization

File: source/core/slang-gcc-compiler-util.cpp [hunk 1]
```diff
@@ -369,7 +369,8 @@ static SlangResult _parseGCCFamilyLine(const UnownedStringSlice& line, LineParse
 
     if (options.sourceType == SourceType::CPP)
     {
-        cmdLine.addArg("-std=c++11");
+        // Need C++14 for partial specialization
+        cmdLine.addArg("-std=c++14");
```
Comment (tfoleyNV) [inline source/core/slang-gcc-compiler-util.cpp]: Ideally our C++ output should be fairly conservative in what it required. I'd think that C++11 would be a fairly high version to require, and would prefer that we make things work with C++98 if at all possible.

Obviously if we aspire to have C output work some day, then we'd need to reduce our reliance on fancy C++ features to near-zero anyway.

That can be addressed in the long run.

File: tests/cross-compile/c-cross-compile.slang.expected [hunk 2]
```diff
@@ -0,0 +1,110 @@
+result code = 0
```
Comment (tfoleyNV) [inline tests/cross-compile/c-cross-compile.slang.expected]: I'm a little unclear on why we have an HLSL output here as the expected output of `c-cross-compile`.

**Pull request merged**